contracts/stake/src/state.rs

```diff
@@ -70,204 +67,214 @@ impl StakeState {
+        let digest = unstake.signature_message().to_vec();
+        if !rusk_abi::verify_bls(digest, account, signature) {
             panic!("Invalid signature!");
         }
-        // make call to transfer contract to withdraw a note from this contract
-        // containing the value of the stake
-        let withdraw_data = Mint {
-            value,
-            address: unstake.address,
-            sender: unstake.public_key,
-        };
+
+        // make call to the transfer contract to withdraw funds from this
+        // contract into the receiver specified by the withdrawal.
         let _: () =
-            rusk_abi::call(TRANSFER_CONTRACT, "withdraw", &withdraw_data)
-                .expect(
-                "Withdrawing the balance from contract should be successful",
-            );
+            rusk_abi::call(TRANSFER_CONTRACT, "withdraw", transfer_withdraw)
+                .expect("Withdrawing stake should succeed");
+
+        // update the state accordingly
+        loaded_stake.amount = None;
 
         rusk_abi::emit(
             "unstake",
-            StakingEvent {
-                public_key: unstake.public_key,
-                value,
+            StakeEvent {
+                account,
+                value: staked_value,
+                receiver: Some(*transfer_withdraw.receiver()),
             },
         );
 
-        let key = unstake.public_key.to_bytes();
+        let key = account.to_bytes();
         self.previous_block_state
             .entry(key)
-            .or_insert((prev_value, unstake.public_key));
+            .or_insert((prev_stake, account));
     }
 
     pub fn withdraw(&mut self, withdraw: Withdraw) {
-        // deplete the stake from a key and increment the signature counter
-        let loaded_stake = self
-            .get_stake_mut(&withdraw.public_key)
-            .expect("A stake should exist in the map to be withdrawn!");
+        self.check_new_block();
 
-        let counter = loaded_stake.counter();
-        let reward = loaded_stake.reward();
+        let transfer_withdraw = withdraw.transfer_withdraw();
+        let account = *withdraw.account();
+        let value = transfer_withdraw.value();
+        let signature = *withdraw.signature();
 
-        if reward == 0 {
-            panic!("Nothing to withdraw!");
-        }
-
-        loaded_stake.deplete_reward();
-        loaded_stake.increment_counter();
+        let loaded_stake = self
+            .get_stake_mut(&account)
+            .expect("A stake should exist in the map to be unstaked!");
 
-        // verify signature
-        let digest = Withdraw::signature_message(
-            counter,
-            withdraw.address,
-            withdraw.nonce,
-        )
-        .to_vec();
+        // ensure there is a reward, and that the withdrawal is not more than
+        // that
+        if value > loaded_stake.reward {
+            panic!("Value withdrawn larger than available reward");
+        }
```

Comment:
Why the behaviour of this has changed?
Before all the reward was withdrawn, that was simple.

Furthermore:
- I cannot see the benefit of withdraw partial reward
- This was done in order to minimize the note fragmentation

Reply:
Since we now have Moonlight, note fragmentation might not be that much of a concern. However, I probably shouldn't change this in this PR.

Will remove the change.

I can see a benefit though. Doesn't slashing affect the reward amount? Say a user submits a transaction which calls `reward`. Then, before the transaction is accepted into a block, they get slashed. The transaction would then inevitably, further punishing the user in an unintended way. The same applies for stakes, but that is a whole different problem.
Partial withdrawals would allow a user to manually mitigate the risk of panic.

Reply:
Staking is not depleting from user's reward anymore.

Reply:
In the current implementation, slashing either burns part of the stake or it moves it to rewards. So slashing does affect the reward amount but only to increase it.

That said, I tend to agree with Seppia in that I see no value in withdrawing part of the reward. For what is worth, I also think leaving the reward there after the unstake has no value. Which is why I so far assumed that unstaking implied withdrawing the whole stake plus the reward

Reply:
>  For what is worth, I also think leaving the reward there after the unstake has no value

2 Reason:
1. We have to leave the reward there after the unstake in order to prevent the edge case of having an unstake operation in the same block where the provisioner is rewarded
2. Unstake operation is currently withdrawing from contract, WithdrawReward is minting. So it's an implementation detail that we cannot mix those

Reply:
I said it has no value, not that it wasn't necessary :-p
Anyway, both are valid points from an implementation point of view, but still technical details.

For the sake of discussion: 
- Point one could be fixed by having unstake operations take place after rewardings. If instead you refer to the case in which the unstake occurs in the same block for which the provisioner voted (as part of the committee), we could handle the case by forbidding a provisioner to vote for a block including its own unstake transaction.
- Point two could be fixed by minting upon generating the reward, not when withdrawing it. Which btw seems more correct to me, taking into account that we consider that money as part of the contract deposit balance.